apps/anoma_node/lib/node/transaction/ordering.ex

```diff
@@ -357,20 +579,483 @@ defmodule Anoma.Node.Transaction.Ordering do
+  @spec recalculate_and_update_watermarks_for_keys([binary()], t()) :: t()
+  defp recalculate_and_update_watermarks_for_keys(keys, state) do
+    Enum.reduce(keys |> Enum.uniq(), state, fn key, acc_state ->
+      recalculate_and_update_watermarks_for_key(key, acc_state)
+    end)
+  end
+
+  # Recalculates watermarks for a single key and sends updates to shards if they advance.
+  @spec recalculate_and_update_watermarks_for_key(binary(), t()) :: t()
+  defp recalculate_and_update_watermarks_for_key(key, state) do
+    current_wms =
+      Map.get(state.watermark_state, key, %{
+        read_watermark: -1,
+        write_watermark: -1
+      })
+
+    new_read_wm = calculate_watermark(key, :read, state)
+    new_write_wm = calculate_watermark(key, :write, state)
+
+    state_after_update = state
+
+    # Check and update read watermark
+    state_after_update =
+      if new_read_wm > current_wms.read_watermark do
+        case get_shard_pid(key, state) do
+          {:ok, shard_pid} ->
+            send(shard_pid, {:read_watermark_advanced, key, new_read_wm})
+
+          {:error, reason} ->
+            Logger.error(
+              "Failed to get shard PID for key #{inspect(key)} when advancing read watermark: #{inspect(reason)}"
+            )
+
+            # Continue without sending update
+        end
+
+        updated_key_wms = Map.put(current_wms, :read_watermark, new_read_wm)
+
+        %{
+          state_after_update
+          | watermark_state:
+              Map.put(
+                state_after_update.watermark_state,
+                key,
+                updated_key_wms
+              )
+        }
+      else
+        # No change
+        state_after_update
+      end
+
+    # Check and update write watermark
+    # Re-fetch current_wms in case read watermark was updated
+    current_wms_after_read =
+      Map.get(state_after_update.watermark_state, key, %{
+        read_watermark: -1,
+        write_watermark: -1
+      })
+
+    state_after_update =
+      if new_write_wm > current_wms_after_read.write_watermark do
+        case get_shard_pid(key, state) do
+          {:ok, shard_pid} ->
+            send(shard_pid, {:write_watermark_advanced, key, new_write_wm})
+
+          {:error, reason} ->
+            Logger.error(
+              "Failed to get shard PID for key #{inspect(key)} when advancing write watermark: #{inspect(reason)}"
+            )
+
+            # Continue without sending update
+        end
+
+        updated_key_wms =
+          Map.put(current_wms_after_read, :write_watermark, new_write_wm)
+
+        %{
+          state_after_update
+          | watermark_state:
+              Map.put(
+                state_after_update.watermark_state,
+                key,
+                updated_key_wms
+              )
+        }
+      else
+        # No change
+        state_after_update
+      end
+
+    # Return final state for this key
+    state_after_update
```

Comment:
Maybe we can flip the case, so the shorter case comes first, so it's easier to see what block relates to what, I have a hard time intuitively seeing the relation when the shorter case comes second